[HLSL] Support packoffset attribute in AST

python3kgae opened the pull request:
Add HLSLPackOffsetAttr to save packoffset in AST.

Since we have to parse the attribute manually in ParseHLSLAnnotations, we could create the ParsedAttribute with a integer offset parameter instead of string. This approach avoids parsing the string if the offset is saved as a string in HLSLPackOffsetAttr.

For #57914.

python3kgae committed:
[HLSL] Support packoffset attribute in AST

Add HLSLPackOffsetAttr to save packoffset in AST.

Since we have to parse the attribute manually in ParseHLSLAnnotations,
we could create the ParsedAttribute with a integer offset parameter instead of string.
This approach avoids parsing the string if the offset is saved as a string in HLSLPackOffsetAttr.

python3kgae committed:
Fix doc format.

hekota commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -42,6 +42,54 @@ Decl *SemaHLSL::ActOnStartBuffer(Scope *BufferScope, bool CBuffer,
+        VarDecl *OtherVar = Range.first;
+        unsigned OtherBegin = Range.second.first;
+        unsigned OtherEnd = Range.second.second;
+        if (Begin < OtherEnd && OtherBegin < Begin) {
+          Diag(Var->getLocation(), diag::err_hlsl_packoffset_overlap)
+              << Var << OtherVar;
+          break;
+        } else if (OtherBegin < End && Begin < OtherBegin) {
+          Diag(Var->getLocation(), diag::err_hlsl_packoffset_overlap)
+              << Var << OtherVar;
+          break;
+        }
```
I don't think this covers all cases and it can be simplified. We know `Begin` is always less than `End`. The ranges do not overlap when `End <= OtherBegin || Begin >= OtherEnd`. So if I negate this condition I get:

```
if (End > OtherBegin && Begin < OtherEnd) {
  Diag(...)
}
```

hekota commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -42,6 +42,54 @@ Decl *SemaHLSL::ActOnStartBuffer(Scope *BufferScope, bool CBuffer,
+  } else if (HasPackOffset) {
+    ASTContext &Context = getASTContext();
+    // Make sure no overlap in packoffset.
+    llvm::SmallDenseMap<VarDecl *, std::pair<unsigned, unsigned>>
```
Why use a hashtable when no lookup is needed? You can use the SmallVector above to store (Var, Begin, End) when you find the attribute and then iterate over that. No need for 2 data structures.

hekota replied:
You could also sort the ranges (by `Begin`) and then walk through them to make sure they are incrementing. That would avoid the O(n^2) overlap check.

python3kgae committed:
Update per comment.

hekota reviewed (approved):
LGTM!

llvm-beanz commented on clang/test/AST/HLSL/packoffset.hlsl:
```diff
@@ -0,0 +1,16 @@
+    float C2 : packoffset(c1);
+    // CHECK-NEXT: VarDecl {{.*}} col:11 C3 'float'
+    // CHECK-NEXT: HLSLPackOffsetAttr {{.*}} 5
+    float C3 : packoffset(c1.y);
```
How about some valid cases for vector and array cbuffer elements?

python3kgae replied:
Updated.

llvm-beanz commented on clang/test/SemaHLSL/packoffset-invalid.hlsl:
```diff
@@ -0,0 +1,55 @@
+    ST A1 : packoffset(c0.y);
+    // expected-error@+1{{packoffset cannot cross register boundary}}
+    float A2[2] : packoffset(c1.w);
+}
```
Some additional error cases to consider:
```
// Some parsing fails
packoffset();
packoffset((c0));
packoffset(c0, x);
packoffset(c.x);
packoffset;
packoffset(;
packoffset);
packoffset c0.x;

// Some other odd failures that look more real
packoffset(c0.xy);
packoffset(c0.rg);

packoffset(c0.yes);
packoffset(c0.woo);

packoffset(c0.xr);
```

python3kgae replied:
Updated.

llvm-beanz commented on clang/lib/Parse/ParseHLSL.cpp:
```diff
@@ -183,6 +183,86 @@ void Parser::ParseHLSLAnnotations(ParsedAttributes &Attrs,
+    ArgExprs.push_back(IntegerLiteral::Create(
+        Ctx, llvm::APInt(Ctx.getTypeSize(Ctx.getSizeType()), Offset),
+        Ctx.getSizeType(),
+        SourceLocation())); // Dummy location for integer literal.
```
Why not give this the location of the parsed identifier?

python3kgae replied:
Updated.

llvm-beanz commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -7314,6 +7314,41 @@ static void handleHLSLSV_DispatchThreadIDAttr(Sema &S, Decl *D,
+static void handleHLSLPackOffsetAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  if (!isa<VarDecl>(D) || !isa<HLSLBufferDecl>(D->getDeclContext())) {
+    S.Diag(AL.getLoc(), diag::err_hlsl_attr_invalid_ast_node)
+        << AL << "cbuffer constant";
```
Maybe:
```suggestion
        << AL << "cbuffer member variable";
```

"cbuffer constant" is less clear to me because the variable declaration isn't actually marked constant in the source, it is just implicitly constant because it is in a cbuffer.

python3kgae replied:
Updated.

llvm-beanz commented on clang/test/AST/HLSL/packoffset.hlsl:
```diff
@@ -0,0 +1,16 @@
+// CHECK: HLSLBufferDecl {{.*}} cbuffer A
+cbuffer A
+{
+    // CHECK-NEXT: VarDecl {{.*}} C1 'float4'
```
We could use some tests for `double` and `half` scalars, arrays and vectors.

python3kgae replied:
Updated.

python3kgae committed:
Check double alignment.
Support rgba as component.
Add calculateLegacyCbufferSize to get correct size when check overlap.

python3kgae commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7398,6 +7398,26 @@ The full documentation is available here: https://docs.microsoft.com/en-us/windo
+  let Content = [{
+The packoffset attribute is used to change the layout of a cbuffer.
+Attribute spelling in HLSL is: ``packoffset( c[Subcomponent][.component] )``.
+A subcomponent is a register number, which is an integer. A component is in the form of [.xyzw].
```
Updated.

python3kgae committed:
Update error message.

llvm-beanz commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -4372,6 +4372,13 @@ def HLSLResourceBinding: InheritableAttr {
+def HLSLPackOffset: HLSLAnnotationAttr {
+  let Spellings = [HLSLAnnotation<"packoffset">];
+  let LangOpts = [HLSL];
+  let Args = [IntArgument<"Offset">];
```
Looking at this more this seems really weird to me. The argument isn't actually an integer, it's a special string. You're parsing it to an integer, but that actually loses some fidelity.

We should probably store index and sub-component values in the attribute rather than collapsing it down into one value.

python3kgae replied:
Do you mind the value store as integer or string?

python3kgae replied:
Information lost is the difference between 'c0' and 'c',  'xyzw' and 'rgba'.
If we want to keep that, they'll need to be kept these as string.
Then we'll need to parse the string when using the value as integer.

llvm-beanz replied:
I think it is fine to keep these as integers.

llvm-beanz replied:
The feedback here still isn't resolved. The argument is an IdentifierArgument with special parsing handling, it isn't an integer. We can store the parsed integer values in the attribute but we shouldn't be calling the argument an IntArgument.

damyanp reviewed (changes requested):
Apologies, it looks like I forgot to submit this review!  Hopefully the comments aren't too out of date.

python3kgae commented on clang/include/clang/Basic/DiagnosticParseKinds.td:
```diff
@@ -1745,5 +1745,7 @@ def err_hlsl_separate_attr_arg_and_number : Error<"wrong argument format for hls
 def ext_hlsl_access_specifiers : ExtWarn<
   "access specifiers are a clang HLSL extension">,
   InGroup<HLSLExtension>;
+def err_hlsl_unsupported_component : Error<"invalid component '%0' used; expected 'x', 'y', 'z', or 'w'">;
+def err_hlsl_packoffset_invalid_reg : Error<"invalid resource class specifier '%0' for packoffset, expected 'c'">;
```
Yes. I'll add that.

python3kgae committed:
Add the difference for mix packoffset and non-packoffset to the difference doc.

python3kgae commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7398,6 +7398,26 @@ The full documentation is available here: https://docs.microsoft.com/en-us/windo
+Attribute spelling in HLSL is: ``packoffset( c[Subcomponent][.component] )``.
+A subcomponent is a register number, which is an integer. A component is in the form of [.xyzw].
+
+Here're packoffset examples with and without component:
```
Updated.

llvm-beanz commented on clang/docs/HLSL/ExpectedDifferences.rst:
```diff
@@ -108,3 +108,18 @@ behavior between Clang and DXC. Some examples include:
+DXC allows mixing packoffset and non-packoffset elements in the same cbuffer,
+accompanied by a warning.
+
+However, both Clang and FXC do not permit this and instead report an error.
```
I worry about this. I thought we had discussed this and DXC did something sane and reasonable so we could do the same in Clang. Is there some detail I'm missing as to why we can't match DXC.

(cc @tex3d)

llvm-beanz reviewed (changes requested):
I'm marking this as requesting changes because I don't think we should land this as-is.

python3kgae committed:
Change to error on mix into warning.

python3kgae commented:
> I'm marking this as requesting changes because I don't think we should land this as-is.

Switched to warning.

python3kgae committed:
Switch to save 2 int args.

python3kgae committed:
Remove the expected difference.

python3kgae committed:
Add warning group for mix packoffset.
